File: tests/r_parity/55_hc2_hc3.py
"""StatsPAI HC2 / HC3 heteroskedasticity-robust SE parity (Python side) -- Module 55.

Companion ``55_hc2_hc3.R`` runs ``lm`` + ``sandwich::vcovHC`` on identical bytes.

Module 01 pins the HC1 heteroskedasticity-robust SE. This module pins the
two small-sample variants referees increasingly prefer for finite samples
(MacKinnon & White 1985):

  * **HC2**: leverage adjustment ``e_i / sqrt(1 - h_ii)``.
  * **HC3**: jackknife-style adjustment ``e_i / (1 - h_ii)`` (more
    conservative; approximates the delete-one jackknife).

``sp.regress(robust="hc2"/"hc3")`` and ``sandwich::vcovHC(type="HC2"/"HC3")``
implement the same MacKinnon-White adjustments, so both rows are a strict
machine-precision headline (rel_se ~ 1e-9). This certifies an additional
``sp.regress`` covariance path; it does not introduce a new certified
symbol (``regress`` is already certified via modules 01/14).
"""
from __future__ import annotations

import statspai as sp

from _common import ParityRecord, dump_csv, write_results


MODULE = "55_hc2_hc3"
FORMULA = "lemp ~ treat + year"


def main() -> None:
    df = sp.datasets.mpdta()
    dump_csv(df, MODULE)

    fits = {"hc2": sp.regress(FORMULA, data=df, robust="hc2"),
            "hc3": sp.regress(FORMULA, data=df, robust="hc3")}

    n = int(fits["hc2"].data_info.get("n_obs", len(df)))
    rows: list[ParityRecord] = []
    for kind, fit in fits.items():
        for name in fit.params.index:
            beta = float(fit.params[name])
            canonical = "(Intercept)" if name == "Intercept" else name
            rows.append(
                ParityRecord(
                    module=MODULE, side="py",
                    statistic=f"{kind}_{canonical}",
                    estimate=beta, se=float(fit.std_errors[name]), n=n,
                )
            )

    write_results(
        MODULE, "py", rows,
        extra={
            "formula": FORMULA,
            "vcov": "HC2 + HC3 (MacKinnon-White 1985)",
            "matches": "sandwich::vcovHC(type='HC2'/'HC3')",
        },
    )


if __name__ == "__main__":
    main()

File: tests/r_parity/55_hc2_hc3.R
# StatsPAI HC2 / HC3 heteroskedasticity-robust SE parity (R side) -- Module 55.
#
# Reads data/55_hc2_hc3.csv and runs lm + sandwich::vcovHC types HC2 and HC3.
# These are the MacKinnon-White (1985) small-sample heteroskedasticity-robust
# variants; sp.regress(robust="hc2"/"hc3") implements the same adjustments,
# so each row must match to machine precision (module 01 covers HC1).

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(sandwich)
})

MODULE  <- "55_hc2_hc3"
FORMULA <- lemp ~ treat + year

df <- read_csv_strict(MODULE)

fit <- lm(FORMULA, data = df)
betas <- coef(fit)

rows <- list()
for (kind in c("hc2", "hc3")) {
  se <- sqrt(diag(vcovHC(fit, type = toupper(kind))))
  for (name in names(betas)) {
    beta <- unname(betas[name])
    s    <- unname(se[name])
    rows[[length(rows) + 1L]] <- parity_row(
      module    = MODULE,
      statistic = paste0(kind, "_", name),
      estimate  = beta,
      se        = s,
      ci_lo     = beta - qnorm(0.975) * s,
      ci_hi     = beta + qnorm(0.975) * s,
      n         = nrow(df)
    )
  }
}

write_results(MODULE, rows,
              extra = list(formula = deparse(FORMULA),
                           vcov = "HC2 + HC3 (sandwich::vcovHC)"))
